core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java

```diff
@@ -299,29 +301,36 @@ private static CatalogHandle requiredCatalogHandle(PartitioningHandle partitioni
                 new IllegalStateException("No catalog handle for partitioning handle: " + partitioningHandle));
     }
 
-    private static List<InternalNode> createArbitraryBucketToNode(long seed, List<InternalNode> nodes, int bucketCount)
+    @VisibleForTesting
+    static List<InternalNode> createArbitraryBucketToNode(long seed, List<InternalNode> nodes, int bucketCount)
     {
+        // Assign each bucket to the machine with the highest weight (hash) and keep the number of buckets per node balanced.
+        // This is Rendezvous Hashing (Highest Random Weight) algorithm with cap on max buckets per node.
         requireNonNull(nodes, "nodes is null");
         checkArgument(!nodes.isEmpty(), "nodes is empty");
         checkArgument(bucketCount > 0, "bucketCount must be greater than zero");
 
-        // Assign each bucket to the machine with the highest weight (hash)
-        // This is simple Rendezvous Hashing (Highest Random Weight) algorithm
+        int maxPerNode = Math.ceilDiv(bucketCount, nodes.size());
+
+        Map<InternalNode, Integer> nodeBucketCounts = new HashMap<>();
+        for (InternalNode node : nodes) {
+            nodeBucketCounts.put(node, 0);
+        }
+
         ImmutableList.Builder<InternalNode> bucketAssignments = ImmutableList.builderWithExpectedSize(bucketCount);
+
         for (int bucket = 0; bucket < bucketCount; bucket++) {
             long bucketHash = XxHash64.hash(seed, bucket);
 
-            InternalNode bestNode = null;
-            long highestWeight = Long.MIN_VALUE;
-            for (InternalNode node : nodes) {
-                long weight = XxHash64.hash(node.longHashCode(), bucketHash);
-                if (weight >= highestWeight) {
-                    highestWeight = weight;
-                    bestNode = node;
-                }
-            }
+            InternalNode node = nodeBucketCounts.keySet().stream()
+                    .max(Comparator.comparingLong(n -> XxHash64.hash(n.longHashCode(), bucketHash)))
```

Comment:
Just curious what's hash algo stand for now, normal hash is also expected works here?